runtime/lua/vim/net/_ssh.lua

```diff
@@ -234,4 +238,198 @@ function M.get_hosts()
+local processed_idx = 0
+local job_id = 0
+local is_remote_running = false
+
+local function _run_cmd(cmd, on_exit)
+  stdout = {}
+  processed_idx = 0
+
+  job_id = vim.fn.jobstart(cmd, {
+    pty = true,
+    on_stdout = function(_job_id, data, _event)
+      data = data ---@type string[]
+      for _, chunk in ipairs(data) do
+        local line = chunk:gsub('\r', '\n')
+        table.insert(stdout, line)
+      end
+
+      local unprocessed_data = table.concat(vim.list_slice(stdout, processed_idx + 1))
+      if unprocessed_data:find('Tunnel created successfully') then
+        is_remote_running = true
+      end
+      for i, prompt in ipairs(ssh_prompts) do
+        if unprocessed_data:find(vim.pesc(prompt.prompt)) then
+          local resp = ''
+          if prompt.value ~= '' then
+            resp = prompt.value
+          else
+            local unprocessed_data_lines = vim.split(vim.trim(unprocessed_data), '\n')
+            local input_label =
+              string.format('%s ', unprocessed_data_lines[#unprocessed_data_lines])
+            if prompt.type == 'secret' then
+              resp = vim.fn.inputsecret(input_label)
+            else
+              resp = vim.fn.input(input_label)
+            end
+            vim.cmd('redraw')
+          end
+
+          ssh_prompts[i].value = resp
+          processed_idx = #stdout
+          vim.api.nvim_chan_send(job_id, resp .. '\n')
+        end
+      end
+    end,
+    on_exit = function(chan_id, data, event)
+      if on_exit ~= nil then
+        on_exit(chan_id, data, event)
+      end
+    end,
+  })
+
+  return job_id
+end
+
+--- Starts a server on the remote machine via ssh and tunnels it to a local socket.
+---@param address string
+---@return string local_socket
+function M.connect_to_address(address)
+  if not address or address == '' then
+    error('invalid SSH address given')
+  end
+
+  if vim.fn.executable('ssh') ~= 1 then
+    error('ssh not installed on system')
+  end
+
+  -- Reset the prompts table
+  for i, prompt in ipairs(ssh_prompts) do
+    if prompt.value ~= '' then
+      ssh_prompts[i].value = ''
+    end
+  end
+
+  local remote_has_nvim = false
+  local check_cmd = string.format('ssh -t %s %s', address, vim.fn.shellescape('nvim --clean -v'))
+  local check_on_exit = function(_job_id, _data, _event)
+    local check_str = table.concat(stdout, '')
+    if check_str:find('NVIM v.*') ~= nil then
+      remote_has_nvim = true
+    end
+  end
+  local check_id = _run_cmd(check_cmd, check_on_exit)
+
+  vim.fn.jobwait({ check_id })
+
+  if not remote_has_nvim then
+    error('Neovim needs to be installed on the remote machine')
+  end
+
+  local local_socket = vim.fn.stdpath('run') .. '/host_nvim_' .. vim.fn.getpid() .. '.pipe'
+  local remote_socket = ''
+
+  local find_socket_cmd = string.format(
+    'ssh -t %s %s',
+    address,
+    vim.fn.shellescape(
+      'nvim --headless --cmd \'echo "Free server: " . v:servername . nr2char(10) | :qall\''
```

Comment:
all the escaping and shellecape() would not be needed if the full command was passed as a list:

    cmd = { 'ssh', '-t', address, "nvim --headless --cmd 'echo..." }